### PR TITLE
Support tenant manifests in the scrubber

### PR DESCRIPTION
**Pull request by arpad-m**
Support tenant manifests in the storage scrubber:

* list the manifests, order them by generation
* delete all manifests except for the two most recent generations
* for the latest manifest: try parsing it.

I've tested this patch by running the  against a staging bucket and it successfully deleted stuff (and avoided deleting the latest two generations).

In follow-up work, we might want to also check some invariants of the manifest, as mentioned in #8088.

Part of #9386
Part of #8088

### CODE REVIEW

File: pageserver/src/tenant/remote_timeline_client.rs [hunk 1]
```diff
@@ -2564,9 +2564,9 @@ pub fn parse_remote_index_path(path: RemotePath) -> Option<Generation> {
 }
 
 /// Given the key of a tenant manifest, parse out the generation number
-pub(crate) fn parse_remote_tenant_manifest_path(path: RemotePath) -> Option<Generation> {
+pub fn parse_remote_tenant_manifest_path(path: RemotePath) -> Option<Generation> {
     static RE: OnceLock<Regex> = OnceLock::new();
-    let re = RE.get_or_init(|| Regex::new(r".+tenant-manifest-([0-9a-f]{8}).json").unwrap());
+    let re = RE.get_or_init(|| Regex::new(r".*tenant-manifest-([0-9a-f]{8}).json").unwrap());
```
Comment (problame) [inline pageserver/src/tenant/remote_timeline_client.rs:2569]: This method is called sometimes with a full key, sometimes with just the basename.
What prompted this `.+` => `.*` change?
If we're drive-by-fixing stuff anyway, can we also ensure we always evaluate the regex against a basename?

File: storage_scrubber/src/checks.rs [hunk 2]
```diff
@@ -527,3 +531,136 @@ async fn list_timeline_blobs_impl(
...
+            },
+        });
+    }
+
+    // Find the manifest with the highest generation
+    let (latest_generation, latest_listing_object) = manifests
+        .iter()
+        .max_by_key(|i| i.0)
+        .map(|(g, obj)| (*g, obj.clone()))
+        .unwrap();
```
Comment (problame) [inline storage_scrubber/src/checks.rs:616]: If there was any error (`!unknown_keys.is_empty()`), then we must not assume we actually see all the manifests.
We should bail with some sort of warning or error instead and not do anything.

File: pageserver/src/tenant/remote_timeline_client.rs (hunk 1 above)
Comment (arpad-m) [reply pageserver/src/tenant/remote_timeline_client.rs:2569]: > This method is called sometimes with a full key, sometimes with just the basename.
> What prompted this .+ => .* change?

That's why i did the `.+` => `.*` change.

File: storage_scrubber/src/pageserver_physical_gc.rs [hunk 3]
```diff
@@ -484,8 +598,41 @@ pub async fn pageserver_physical_gc(
...
-    let timelines = timelines.try_buffered(CONCURRENCY);
+    enum GcSummaryOrContent<T> {
+        Content(T),
+        GcSummary(GcSummary),
+    }
+    let timelines = tenants.map_ok(|tenant_shard_id| {
+        let target_ref = &target;
+        let remote_client_ref = &remote_client;
+        async move {
+            let summaries_from_manifests = match gc_tenant_manifests(
```
[the remark this answers is not in the document]
Comment (arpad-m) [reply storage_scrubber/src/pageserver_physical_gc.rs:609]: In follow-up work I'd like to use tenant manifest data to check invariants, like e.g. that timelines that are offloaded are also marked as archived. That needs timelines to access tenant specific data.

Also I'm not sure how to cheaply do the `tee` operation: caching all to RAM once is not an option, and cloning the stream (if that's even possible) would, from what I can tell, do the listing operations twice, so it would mean more network operations.

File: storage_scrubber/src/checks.rs [hunk 4]
```diff
@@ -527,3 +531,136 @@ async fn list_timeline_blobs_impl(
...
+            };
+            if !name.starts_with("tenant-manifest") {
+                break 'err;
+            }
+            let Ok(path) = RemotePath::from_string(name) else {
+                break 'err;
+            };
+            let Some(generation) = parse_remote_tenant_manifest_path(path) else {
+                break 'err;
+            };
```
[the remark this answers is not in the document]
Comment (arpad-m) [reply storage_scrubber/src/checks.rs:589]: good idea, done

File: storage_scrubber/src/checks.rs [hunk 5]
```diff
@@ -527,3 +531,136 @@ async fn list_timeline_blobs_impl(
...
+                manifest_info: RemoteTenantManifestInfo {
+                    latest_generation: Some(latest_generation),
+                    manifests,
+                    unknown_keys,
+                },
+            });
+        }
+        Err(parse_error) => errors.push(format!(
+            "tenant-manifest.json body parsing error: {parse_error}"
+        )),
```
[the remark this answers is not in the document]
Comment (arpad-m) [reply storage_scrubber/src/checks.rs:649]: there could theoretically be multiple errors per key, so a `Vec<(RemotePath, String)>` is better. In general, we shouldn't spend too much time on this because it's "just" logged later on anyways, where we include the tenant id.

[the remark this answers is not in the document]
Comment (arpad-m) [reply storage_scrubber/src/checks.rs:649]: I mean, maybe it's useful for later refactors to put it into some kind of report that we put into the storcon db.

File: storage_scrubber/src/pageserver_physical_gc.rs [hunk 6]
```diff
@@ -451,13 +521,56 @@ async fn gc_ancestor(
...
+        mut manifest_info,
+    } = list_tenant_manifests(remote_client, tenant_shard_id, target).await?;
+    if !errors.is_empty() {
+        for error in errors {
+            tracing::warn!(%tenant_shard_id, "list_tenant_manifests: {error}");
+        }
+    } else {
+        let Some(latest_gen) = manifest_info.latest_generation else {
+            return Ok(gc_summary);
+        };
```
[the remark this answers is not in the document]
Comment (arpad-m) [reply storage_scrubber/src/pageserver_physical_gc.rs:543]: yeah that would make sense.

File: storage_scrubber/src/checks.rs (hunk 2 above)
Comment (arpad-m) [reply storage_scrubber/src/checks.rs:617]: done

File: storage_scrubber/src/pageserver_physical_gc.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (arpad-m) [reply storage_scrubber/src/pageserver_physical_gc.rs]: addressed

File: storage_scrubber/src/checks.rs [hunk 7]
```diff
@@ -527,3 +531,132 @@ async fn list_timeline_blobs_impl(
...
+            // TODO a let chain would be nicer here.
+            let Some(name) = key.object_name() else {
+                break 'err;
+            };
+            if !name.starts_with(TENANT_MANIFEST_STEM) {
+                break 'err;
+            }
+            let Some(generation) = parse_remote_tenant_manifest_path(key.clone()) else {
+                break 'err;
+            };
```
Comment (problame) [inline storage_scrubber/src/checks.rs:586]: the check for `starts_with` and `parse_remote_tenant_manifest_path` still seems redundant

File: storage_scrubber/src/pageserver_physical_gc.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (problame) [reply storage_scrubber/src/pageserver_physical_gc.rs:611]: Hm, fair points.

But still, this GcSummaryOrContent constructs reads very awkwardly.

Maybe the issue is that we're not constructing the entire stream pipeline in one go, but have this chunk here, then define `gc_timeline` below, then continue consuming the stream.

Might make sense to move `gc_timeline` out of this function body, then use sth like `flatten` or `fuse`?

File: storage_scrubber/src/checks.rs (hunk 7 above)
Comment (arpad-m) [reply storage_scrubber/src/checks.rs:586]: > IMO this PR should take the opportunity and consistently evaluate `RemotePath::object_name()` against this regex. In that case, we can start it with `^tenant-manifest-...`

Right above `parse_remote_tenant_manifest_path` there is another parsing function and it takes in the whole path. It would be weird IMO if the two functions worked differently.

The `starts_with` isn't redundant because then it would accept paths like `foobar/absolutely-not-tenant-manifest.json`. If the regex is changed to use `^` it wouldn't be redundant but as per above we should not do that IMO.

File: storage_scrubber/src/pageserver_physical_gc.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (arpad-m) [reply storage_scrubber/src/pageserver_physical_gc.rs:611]: We already do try_flatten below, but moving gc_timeline out of the function body is a sensible change. It bloats the diff a little (mixing code changes that only move stuff with code changes that modify functionality) but I think it's still an okay deal overall given the function is small.

[the remark this answers is not in the document]
Comment (arpad-m) [reply storage_scrubber/src/pageserver_physical_gc.rs]: done the gc_timeline move

File: storage_scrubber/src/checks.rs (hunk 7 above)
Comment (problame) [reply storage_scrubber/src/checks.rs:586]: 🤷

**Review by problame — APPROVED** [review]
